Allow S3 urls in requirements.yml files

dougluce opened the pull request:
##### SUMMARY

Provide a means of grabbing roles from S3 files. Without this capability, there's no clean way to put a packaged role file behind an authentication system.

##### ISSUE TYPE

- Feature Pull Request

##### COMPONENT NAME

ansible-galaxy

##### ANSIBLE VERSION

```
ansible 2.0.1.0
  config file = /Users/doug/.ansible.cfg
  configured module search path = Default w/o overrides
```

dougluce committed:
Allow s3 urls in requirements.yml files

This provide a means of grabbing roles from S3 files. Without this
capability, there's no clean way to put a packaged role file behind an
authentication system.

jimi-c commented on lib/ansible/galaxy/role.py:
```diff
@@ -168,12 +179,24 @@ def fetch(self, role_data):
-                    temp_file.write(data)
+                if archive_url.startswith('s3:'):
+                    bucket, prefix = parse_s3_url(archive_url)
+                    import boto3
```
This needs to be handled better. This import should be at the top of the file and wrapped in a try/except so that boto isn't a hard requirement on ansible-galaxy. Also, this should support both boto and boto3, as both are still in common usage.

dougluce replied:
Just added the changes for that. Looks ugly, but should work.

dougluce committed:
Hoist boto3 import, support boto secondarily

Also avoid making boto/boto3 a requirement for ansible-galaxy in the
general case.

willthames commented on lib/ansible/galaxy/role.py:
```diff
@@ -22,6 +22,17 @@
+
+import re
+
+def parse_s3_url(url):
```
this shouldn't be in the middle of a bunch of imports

dougluce replied:
Yeah, that's a weird place. I didn't intend for it to be right there.

willthames commented on lib/ansible/galaxy/role.py:
```diff
@@ -30,6 +41,14 @@
+try:
+    import boto3
+except ImportError:
+    try:
```
set HAS_BOTO3 before here, so that you can test if you have it later on.

dougluce replied:
What would the test be for?

willthames replied:
The [module checklist](http://docs.ansible.com/ansible/developing_modules.html#module-checklist) (and I know this isn't a module) suggest using 

```
try:
    import boto
    HAS_BOTO = True
except ImportError:
    HAS_BOTO = False
```

With a similar block for `boto3`, you can just use `HAS_BOTO`/`HAS_BOTO3` rather than `try/except NameError`. 

But in this case it's a stylistic thing, and using common Ansible practices, not really a blocker.

willthames replied:
I should have probably used the word 'check' rather than 'test'.

allenluce replied:
Yeah, it'd take a couple of those vars due to boto+boto3 support, and not having to check a flag makes the actual usage faster (assuming that boto3 is favored over boto).

dougluce replied:
Yeah, it'd take a couple of those vars due to boto+boto3 support, and not having to check a flag makes the actual usage faster (assuming that boto3 is favored over boto).

dougluce committed:
Get the parse_s3_url out of the middle of imports

No clue how it ended up there in the first place.

dougluce commented:
I've rebased this to remove conflicts.

jimi-c commented:
@dougluce sorry this didn't make it into 2.1. I've bumped the milestone to 2.2 so we can consider it for the next release.

allenluce commented:
No worries @jimi-c. Anything I can do to make sure it's all ready for inclusion?

dougluce commented:
Looks like this didn't quite make it into 2.2.0. Is there a new milestone for this? Or did something else replace it?

dougluce commented:
Rebased.

mattclay commented on lib/ansible/galaxy/role.py:
```diff
@@ -42,6 +51,13 @@
+    if match:
+        return match.group(1), match.group(2)
+    return None, None
+    
```
CI failure due to PEP 8 issue:

```
2017-01-28 08:02:12 ERROR: PEP 8: lib/ansible/galaxy/role.py:60:1: W293 blank line contains whitespace (legacy)
```

dougluce committed:
Remove blank line

This got nailed by the CI's PEP 8 compliance mechanism.

dougluce commented:
Does anything further need to happen to this to get it merged?

mattclay commented:
Closing and re-opening to trigger CI.

dougluce committed:
PEP8ification

Per CI autocheck.

calfonso commented:
@dougluce We are having further discussion on what direction we are taking Ansible Galaxy roadmap in the near future. We appreciate your patience while we consider your PR.

DanyC97 commented:
@calfonso any notes as to which direction is heading to ?

calfonso commented:
@DanyC97 not just yet, we've been putting off the discussion a bit while we prepared our 2.4 release. This is on our agenda though.

jctanner commented:
bot_skip

sivel commented:
We have discussed this and decided that this is not something we wish to include.  Our recommendation would be to expose the bucket for standard HTTP requests, or to mount the S3 bucket locally and read from the filesystem.

If you have further questions please stop by IRC or the mailing list:

   * IRC: #ansible on irc.freenode.net
   * mailing list: https://groups.google.com/forum/#!forum/ansible-project
